libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -286,16 +294,69 @@ string IRGenerator::constructorCode(ContractDefinition const& _contract)
 
 			<constructorName>(<params>)
 		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.functionName(*constructor));
-
-		out << t.render();
+	t("object", creationObjectName(dynamic_cast<ContractDefinition const&>(*_constructor.scope())));
+	t("allocate", m_utils.allocationFunction());
+	t("assignToParams", params.empty() ? "" : "let " + params + " := ");
+	t("abiDecode", abiFunctions.tupleDecoder(_constructor.functionType(false)->parameterTypes(), true));
+	t("constructorName", m_context.functionName(_constructor));
+	t("params", params);
+
+	return t.render();
+}
+
+std::string IRGenerator::baseConstructorCall(FunctionDefinition const& _baseConstructor)
+{
+	FunctionType constructorType(_baseConstructor);
+	string params;
+	IRGeneratorForStatements generator(m_context, m_utils);
+	if (!constructorType.parameterTypes().empty())
+	{
+		solAssert(m_baseArguments, "");
+		solAssert(m_baseArguments->count(&_baseConstructor), "");
+		std::vector<ASTPointer<Expression>> const* arguments = nullptr;
+		ASTNode const* baseArgumentNode = m_baseArguments->at(&_baseConstructor);
+		if (auto inheritanceSpecifier = dynamic_cast<InheritanceSpecifier const*>(baseArgumentNode))
+			arguments = inheritanceSpecifier->arguments();
+		else if (auto modifierInvocation = dynamic_cast<ModifierInvocation const*>(baseArgumentNode))
+			arguments = modifierInvocation->arguments();
+		solAssert(arguments, "");
+		solAssert(arguments->size() == constructorType.parameterTypes().size(), "");
+		for (unsigned i = 0; i < arguments->size(); ++i)
+			params += (params.empty() ? "" : ", ") +
+				generator.evaluateExpression(
+					*(arguments->at(i)),
+					*constructorType.parameterTypes()[i]
+				).commaSeparatedList();
 	}
 
+	Whiskers t(R"X(
+		<baseArgumentGenerator>
+		<constructorName>(<params>)
+	)X");
+	t("baseArgumentGenerator", generator.code());
+	t("constructorName", m_context.functionName(_baseConstructor));
+	t("params", params);
+
+	return t.render();
+}
+
+string IRGenerator::constructorCode(ContractDefinition const& _contract)
+{
+	// Initialization of state variables in base-to-derived order.
+	solAssert(!_contract.isLibrary(), "Tried to initialize state variables of library.");
+
+	m_baseArguments = &_contract.annotation().baseConstructorArguments;
+
+	ostringstream out;
+	out << stateVariablesCode(_contract);
+
+	if (FunctionDefinition const* constructor = _contract.constructor())
+		out << addContractConstructor(*constructor);
+	else if (auto c = _contract.nextConstructor(m_context.mostDerivedContract()))
+		out << baseConstructorCall(*c);
```

Comment:
Both `c` and `constructor` have the same type but it's harder to realize this if their declarations are different.

I'd suggest either:
```suggestion
	if (auto constructor = _contract.constructor())
		out << addContractConstructor(*constructor);
	else if (auto constructor = _contract.nextConstructor(m_context.mostDerivedContract()))
		out << baseConstructorCall(*constructor);
```

or

```suggestion
	if (FunctionDefinition const* constructor = _contract.constructor())
		out << addContractConstructor(*constructor);
	else if (FunctionDefinition const* constructor = _contract.nextConstructor(m_context.mostDerivedContract()))
		out << baseConstructorCall(*constructor);
```

Reply:
I would stick with FunctionDefinition const* as type is not obvious from expression itself

Reply:
Please try to avoid `auto` as much as possible.

Reply:
but, constructor from second if would shadow previous definition, so going with nextConstructor. Options are
```
if (FunctionDefinition const* constructor = _contract.constructor())
	out << addContractConstructor(*constructor);
else if (FunctionDefinition const* nextConstructor = _contract.nextConstructor(m_context.mostDerivedContract()))
	out << baseConstructorCall(*nextConstructor);
```
or
```
if (FunctionDefinition const* constructor = _contract.constructor())
	out << addContractConstructor(*constructor);
else if ((constructor = _contract.nextConstructor(m_context.mostDerivedContract())))
	out << baseConstructorCall(*constructor);
```

Reply:
Please use different names for different variables :)

Reply:
Did like that at the end

Reply:
> but, constructor from second if would shadow previous definition

Right, my bad. I missed the fact that the first variable is still in the scope of `else`.

`nextConstructor` is good IMO. It reads much better than `c`.